### PR TITLE
[SPARK-47431][SQL][FOLLOWUP] Do not access SQLConf in Literal.apply

### PR DESCRIPTION
**Pull request by mihailomilosevic2001**
### What changes were proposed in this pull request?
Refactor of `literals.scala`. 

### Why are the changes needed?
`Literal` expression is too low-level and should not access `SQLConf`, otherwise it's hard to reason about when/where the default collation is applied.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Tests already available from initial PR.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala [hunk 1]
```diff
@@ -485,7 +484,7 @@ case class Literal (value: Any, dataType: DataType) extends LeafExpression {
   override def sql: String = (value, dataType) match {
     case (_, NullType | _: ArrayType | _: MapType | _: StructType) if value == null => "NULL"
     case _ if value == null => s"CAST(NULL AS ${dataType.sql})"
-    case (v: UTF8String, _: StringType) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala:488]: I don't think we should revert this change?

Comment (mihailom-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala:488]: Added a change for _: StringType, I would only say to wait until implicit casting goes in first, to resolve conflicts on this PR for some more tests that were added and that will be affected by this change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala [hunk 2]
```diff
@@ -485,9 +484,13 @@ case class Literal (value: Any, dataType: DataType) extends LeafExpression {
...
+    case (v: UTF8String, StringType) =>
       // Escapes all backslashes and single quotes.
       "'" + v.toString.replace("\\", "\\\\").replace("'", "\\'") + "'"
+    case (v: UTF8String, st: StringType) =>
+      // Escapes all backslashes and single quotes.
+      "'" + v.toString.replace("\\", "\\\\").replace("'", "\\'") +
+        "'" + st.typeName.substring(6)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala:493]: what does this produce?

Comment (mihailom-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala:493]: 'stringName' COLLATE stringCollation

**Comment by cloud-fan** [comment]
thanks, merging to master!
